protocols/luci-proto-yggdrasil/htdocs/luci-static/resources/protocol/yggdrasil.js

```diff
@@ -0,0 +1,272 @@
+	};
+	if (!value.match(/^([0-9a-fA-F]){128}$/)) {
+		if (value != "auto") {
+			return _('Invalid private key string');
+		}
+		return true;
+	}
+	return true;
+};
+
+function validatePublicKey(section_id,value) {
+	if (value.length == 0) {
+		return true;
+	};
+	if (!value.match(/^([0-9a-fA-F]){64}$/))
+		return _('Invalid public key string');
+	return true;
+};
+
+function validateYggdrasilListenUri(section_id,value) {
+	if (value.length == 0) {
+		return true;
+	};
+	if (!value.match(/^(tls|tcp|unix|quic):\/\//))
+		return _('URI scheme not supported');
+	return true;
+};
+
+function validateYggdrasilPeerUri(section_id,value) {
+	if (!value.match(/^(tls|tcp|unix|quic|socks|socktls):\/\//))
+		return _('URI scheme not supported');
+	return true;
+};
+
+var cbiKeyPairGenerate = form.DummyValue.extend({
+	cfgvalue: function(section_id, value) {
+		return E('button', {
+			'class':'btn',
+			'click':ui.createHandlerFn(this, function(section_id,ev) {
+				var prv = this.section.getUIElement(section_id,'private_key'),
+					pub = this.section.getUIElement(section_id,'public_key'),
+					map = this.map;
+				if ((prv.getValue()||pub.getValue()) && !confirm(_('Do you want to replace the current keys?')))
+					return;
+				return generateKey().then(function(keypair){prv.setValue(keypair.priv);
+					pub.setValue(keypair.pub);
+					map.save(null,true);
+				});
+			},section_id)
+		},[_('Generate new key pair')]);
+	}
+});
+
+function updateActivePeers(ifname) {
+	getPeers(ifname).then(function(peers){
+		var table = document.querySelector('#yggdrasil-active-peerings-' + ifname);
+		if (table) {
+			while (table.rows.length > 1) { table.deleteRow(1); }
+			peers.forEach(function(peer) {
+				var row = table.insertRow(-1);
+				row.style.fontSize = "xx-small";
+				if (!peer.up) {
+					row.style.opacity = "66%";
+				}
+				var cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = peer.remote;
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = peer.up ? "Up" : "Down";
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = peer.inbound ? "In" : "Out";
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.innerHTML = "<u style='cursor: default'>" + peer.address + "</u>"
+				cell.dataToggle = "tooltip";
+				cell.title = "Key: " + peer.key;
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = '%t'.format(peer.uptime);
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = '%.2mB'.format(peer.bytes_recvd);
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = '%.2mB'.format(peer.bytes_sent);
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				cell.textContent = peer.priority;
+
+				cell = row.insertCell(-1)
+				cell.className = "td"
+				if (!peer.up) {
+					cell.innerHTML = "<u style='cursor: default'>%t ago</u>".format(peer.last_error_time)
```

Comment:
Does this produce weird strings? Last error time sounds like an absolute reference, whereas "ago" suffixes an amount (a relative reference). What does ygg output?

Reply:
nothing else to do here beyond making a merge request in yggdrasil for this time edgecase... i will try to reproduce this later.

the 'last error' description text is useful and shouldnt be hard to find. this will need more testing. I do not believe this should block us going forward.